File: src/826.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        int retVal = 0;

        vector<pair<int, int>> job;
        int jobIdx = 0;
        int jobSize = difficulty.size();
        for (int i = 0; i < jobSize; ++i) {
            job.emplace_back(difficulty[i], profit[i]);
        }
        sort(job.begin(), job.end());

        int maxProfit = 0;
        sort(worker.begin(), worker.end());
        for (auto pWork : worker) {
            while ((jobIdx < jobSize) && (pWork >= job[jobIdx].first)) {
                maxProfit = max(maxProfit, job[jobIdx].second);
                ++jobIdx;
            }
            retVal += maxProfit;
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> difficulty;
        vector<int> profit;
        vector<int> worker;
    };
    vector<subject> testData{{{2, 4, 6, 8, 10}, {10, 20, 30, 40, 50}, {4, 5, 6, 7}},
                             {{85, 47, 57}, {24, 66, 99}, {40, 25, 25}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: difficulty = [2,4,6,8,10], profit = [10,20,30,40,50], worker = [4,5,6,7]
     *  Output: 100
     *
     *  Input: difficulty = [85,47,57], profit = [24,66,99], worker = [40,25,25]
     *  Output: 0
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: difficulty = [";
        for (long unsigned int j = 0; j < testData[i].difficulty.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].difficulty[j];
        }
        cout << "], profit = [";
        for (long unsigned int j = 0; j < testData[i].profit.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].profit[j];
        }
        cout << "], worker = [";
        for (long unsigned int j = 0; j < testData[i].worker.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].worker[j];
        }
        cout << "]\n";

        answer = cSolution.maxProfitAssignment(testData[i].difficulty, testData[i].profit, testData[i].worker);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/826.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        retVal = 0

        job = sorted(zip(difficulty, profit))
        jobSize = len(job)
        jobIdx = 0

        maxProfit = 0
        worker.sort()
        for pWorker in worker:
            while (jobIdx < jobSize) and (pWorker >= job[jobIdx][0]):
                maxProfit = max(maxProfit, job[jobIdx][1])
                jobIdx += 1

            retVal += maxProfit

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for difficulty, profit, worker in zip([[2, 4, 6, 8, 10], [85, 47, 57]],
                                              [[10, 20, 30, 40, 50], [24, 66, 99]],
                                              [[4, 5, 6, 7], [40, 25, 25]]):
            # /* Example
            #  *  Input: difficulty = [2,4,6,8,10], profit = [10,20,30,40,50], worker = [4,5,6,7]
            #  *  Output: 100
            #  *
            #  *  Input: difficulty = [85,47,57], profit = [24,66,99], worker = [40,25,25]
            #  *  Output: 0
            #  */
            logging.info(
                "Input: difficulty = %s, profit = %s, worker = %s", difficulty, profit, worker)

            retVal = pSolution.maxProfitAssignment(difficulty, profit, worker)
            logging.info("Output: %d", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
